lib/dsl/lexer.js

```diff
@@ -23,8 +23,51 @@ const Lexer = chevrotain.Lexer;
 
 const tokens = {};
 
+const baseNamePattern = 'baseName';
 const namePattern = /[a-zA-Z_][a-zA-Z_\-\d]*/;
-// All the names tokens of the pegjs implementation have been merged into
+// using start anchor because we are matching with this regexp directly in a custom pattern.
+const namePatternWithStartDigit = /^[\da-zA-Z_][a-zA-Z_\-\d]*/;
+
+/**
+ * Accepts an array of array representing character ranges
+ * e.g: [['a', 'd'], ['0', '2']]
+ *
+ * And returns a set of the "expanded" ranges in charCodes values.
+ * e.g: [97, 99, 98 , 100, 48, 49, 50]
+ *   - 'a'.charCodeAt(0) === 97
+ *   - 'a'.charCodeAt(0) === 100
+ */
+function getRanges(charRanges) {
+  let ranges = [];
+  charRanges.forEach((currRange) => {
+    const startCode = currRange[0].charCodeAt(0);
+    // +1 because lodash range  excludes the end.
+    const endCode = currRange[1].charCodeAt(0) + 1;
+
+    ranges = ranges.concat(_.range(startCode, endCode));
+  });
+
+  return _.uniq(ranges);
+}
+
+// Special NAME Token which follows a baseName config value.
+// It allows starting with numbers or even being only digits.
+const NAME_CAN_START_DIGIT = chevrotain.createToken({
+  name: 'NAME_CAN_START_DIGIT',
+  pattern: (text, offset, matchedTokens) => {
+    const lastToken = _.last(matchedTokens);
+    if (lastToken && lastToken.image === baseNamePattern) {
+      return namePatternWithStartDigit.exec(text.substring(offset));
+    }
+
+    return null;
+  },
+  // this optional property enables performance optimizations in the Lexer.
+  start_chars_hint: getRanges([['0', '9'], ['a', 'z'], ['A', 'Z'], ['_', '_']]),
```

Comment:
If there are inputs like `['_', '_']`, maybe not iterating over it is better by passing a string and checking with `Array::isArray`

Reply:
I don't understand the comment can you demonstrate?

Reply:
```javascript
getRanges([
  ['0', '9'],
  ['a', 'z'],
  ['A', 'Z'],
  '_'
]);
```

Reply:
Yes this version is nicer.